File: 2810/python3/main.py
from collections import deque


class Solution:
    def finalString(self, s: str) -> str:
        arr = deque()
        appendToHead = False
        for c in s:
            if c == "i":
                appendToHead = not appendToHead
            else:
                if appendToHead:
                    arr.appendleft(c)
                else:
                    arr.append(c)
        if appendToHead:
            arr.reverse()
        return "".join(arr)


s = Solution()
print(s.finalString('string'))

File: 2810/go/main.go
package main

// code
func finalString(s string) string {
	arr := []rune{}
	appendToHead := false

	for _, c := range s {
		if c == 'i' {
			appendToHead = !appendToHead
		} else {
			if appendToHead {
                // 创建一个单个元素的 c 的切片
                // 再把原来的切片内的放进去
                // 等价于在前面添加
				arr = append([]rune{c}, arr...)
			} else {
				arr = append(arr, c)
			}
		}
	}
	if appendToHead {
		reverse(arr)
	}
	return string(arr)
}

func reverse(arr []rune) {
	l := len(arr) - 1
	for i, j := 0, l; i < j; i, j = i+1, j-1 {
		arr[i], arr[j] = arr[j], arr[i]
	}
}
